python/nav/mibs/ibm_pdu_mib.py

```diff
@@ -0,0 +1,128 @@
+#
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from nav.mibs import reduce_index
+from nav.mibs.mibretriever import MibRetriever
+from nav.models.manage import Sensor
+
+PHASE_LAST_POWER_READING = 'ibmPduPhaseLastPowerReading'
+
+OUTLET_NAME = 'ibmPduOutletName'
+OUTLET_DESCRIPTION = 'ibmPduOutletDescription'
+OUTLET_VOLTAGE = 'ibmPduOutletVoltage'
+OUTLET_CURRENT = 'ibmPduOutletCurrent'
+OUTLET_MAX_CAPACITY = 'ibmPduOutletMaxCapacity'
+OUTLET_LAST_POWER_READING = 'ibmPduOutletLastPowerReading'
+
+
+class IbmPduMib(MibRetriever):
+    """ Custom class for retrieveing sensors from APC UPSes."""
+    from nav.smidumps.ibm_pdu_mib import MIB as mib
+
+    @inlineCallbacks
+    def get_all_sensors(self):
+        phases = yield self._get_phase_sensors()
+        outlets = yield self._get_outlet_sensors()
+        returnValue(phases + outlets)
+
+    @inlineCallbacks
+    def _get_phase_sensors(self):
+        phases = yield self.retrieve_columns(
+            [PHASE_LAST_POWER_READING]).addCallback(reduce_index)
+        if phases:
+            self._logger.debug("Got phase power readings: %r", phases)
+
+        result = []
+        column = self.nodes.get(PHASE_LAST_POWER_READING, None)
+        for index, row in phases.items():
+            value_oid = str(column.oid + str(index))
+            if len(index) == 2:
+                index = index[0]  # PDU agent breaks the MIB definition :P
+
+            name = "Phase %d" % index
+
+            result.append(dict(
+                oid=value_oid,
+                unit_of_measurement=Sensor.UNIT_WATTS,
+                precision=0,
+                scale=None,
+                description='%s power reading' % name,
+                name=name,
+                internal_name='%s_%s' % (PHASE_LAST_POWER_READING, index),
+                mib=self.get_module_name()
+            ))
+
+        returnValue(result)
+
+    @inlineCallbacks
+    def _get_outlet_sensors(self):
+        outlets = yield self.retrieve_columns([OUTLET_NAME,
+                                               OUTLET_DESCRIPTION,
+                                               OUTLET_VOLTAGE,
+                                               OUTLET_CURRENT,
+                                               OUTLET_MAX_CAPACITY,
+                                               OUTLET_LAST_POWER_READING
+                                               ]).addCallback(reduce_index)
+        if outlets:
+            self._logger.debug("Got outlet power readings: %r", outlets)
+
+        result = []
+        for index, row in outlets.items():
+            for sensor in self._outlet_row_to_sensors(index, row):
+                result.append(sensor)
+        returnValue(result)
+
+    def _outlet_row_to_sensors(self, index, row):
+        name = row.get(OUTLET_NAME)
+        description = row.get(OUTLET_DESCRIPTION)
+
+        voltage = dict(
+            oid=str(self.nodes.get(OUTLET_VOLTAGE, None).oid + str(index)),
```

Comment:
None.oid will not work. If the key is missing it makes more sense to get a KeyError than an AttributeError